### PR TITLE
DS-3636 Indexes and caching performance updates

### PR DESCRIPTION
**Pull request by jonas-atmire**
Pull Request for JIRA ticket:
https://jira.duraspace.org/browse/DS-3636

Put indexes on several widely searched columns + use some additional caching for performance improvements.

### CODE REVIEW

File: dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V5.7_2017.07.04__DS-3636_General_Index_Performance_Updates.sql
Comment (tomdesair) [inline dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V5.7_2017.07.04__DS-3636_General_Index_Performance_Updates.sql]: Please add a drop if exists for "resourcepolicy_type_id_idx"

File: dspace-api/src/main/java/org/dspace/authorize/ResourcePolicy.java
Comment (tomdesair) [inline dspace-api/src/main/java/org/dspace/authorize/ResourcePolicy.java]: Please specify in the code comments why this is more efficient than just creating the record first and then setting all the field values (-> requires only insert statement vs  1 insert + 1 update).

**Review by tomdesair — CHANGES_REQUESTED** [review]
I think this is a no-brainer. I only have a few small remarks.

File: dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V5.7_2017.07.04__DS-3636_General_Index_Performance_Updates.sql
Comment (tomdesair) [inline dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/h2/V5.7_2017.07.04__DS-3636_General_Index_Performance_Updates.sql]: There is already an index on metadata_field_id, so maybe we should just index "place".

**Comment by jonas-atmire** [comment]
Updated the indexes and code comments based on the feedback
Additionally, I noticed that the oracle flyway broke for both this and the previous flyway file
Fixed the following,
- Added slashes between index checking declarations
- Updated some of the index names (These are required to be under 30 characters in Oracle)

**Comment by alanorth** [comment]
This pull request is a no-brainer for me as well. It greatly increases the performance of the XMLUI `/submissions` page in DSpace 5.5, for example. Before this, I observed hundreds or even thousands of locks on the `resourcepolicy` and `metadatavalue` tables during operations on that page. Deleting unfinished submissions would often take several minutes on beefy hardware with modern PostgreSQL 9.5, but complete almost instantly with these indexes.

File: dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/oracle/V5.7_2017.04.11__DS-3563_Index_metadatavalue_resource_type_id_column.sql
Comment (tdonohue) [inline dspace-api/src/main/resources/org/dspace/storage/rdbms/sqlmigration/oracle/V5.7_2017.04.11__DS-3563_Index_metadatavalue_resource_type_id_column.sql]: Is this change required to this old Migration script?  Modifying a previous migration script, even by a single character will change the checksum of that script.  That means that Flyway will no longer be able to validate this migration on databases that previously ran it.  So, with this one character change, anyone running 5.7 or 5.8 will be *required* to run a `dspace database repair` in order to upgrade successfully.  See https://wiki.duraspace.org/display/DSDOC6x/Database+Utilities

**Review by tdonohue — CHANGES_REQUESTED** [review]
Mostly OK, but I need clarification on why we are modifying an old migration script. See inline comments.

**Comment by tdonohue** [comment]
@tomdesair : just a reminder for you to dismiss your "requested changes" review if you are now satisfied with this PR.  It looks to me like the changes you requested have now been made, but I hesitate to dismiss the review of any other Committer.

**Review by tomdesair — APPROVED** [review]

**Comment by J4bbi** [comment]
@kshepherd, should we consider this for 6.4? I haven't tried it yet but it seems relatively straightforward

if something along similar lines hasn't already been applied, it might be worth porting to 7? @tdonohue

**Comment by kshepherd** [comment]
> @kshepherd, should we consider this for 6.4? I haven't tried it yet but it seems relatively straightforward
> 
> if something along similar lines hasn't already been applied, it might be worth porting to 7? @tdonohue

It looks good - is there a dspace 6 port?

**Comment by J4bbi** [comment]
Yeah, #1792 ...... but like I say, haven't tested this myself....

**Pull request closed without merging**
